# 🔲 Chi-Square Test in Statistics
### Testing Association Between Categorical Variables
---

> **Chi-Square Test** is used to determine whether there is a **significant association between two categorical variables**.  
> Unlike Z-Test or T-Test which work on **numerical means**, Chi-Square works on **counts and frequencies**.

---

## 🤔 When to Use Chi-Square Test?

| Test | Data Type | Comparing |
|---|---|---|
| Z-Test | Numerical | Sample mean vs Population mean |
| T-Test | Numerical | Means of groups |
| **Chi-Square Test** | **Categorical** | **Association between two categories** |

---

## 📌 Problem Statement

We use the famous **Titanic Dataset** to answer:

> **"Is there a significant relationship between a passenger's Gender and their Survival?"**

- **Variable 1:** Sex (male / female) → Categorical  
- **Variable 2:** Survived (0 = No, 1 = Yes) → Categorical  

---

## 🔢 Formula

$$\chi^2 = \sum \frac{(O - E)^2}{E}$$

- $O$ = Observed Frequency  
- $E$ = Expected Frequency  
- Higher $\chi^2$ = stronger evidence against H₀

---

## 🧪 Our Hypotheses

- **H₀ (Null Hypothesis):** Gender and Survival are **independent** *(no relationship)*
- **H₁ (Alternative Hypothesis):** Gender and Survival are **dependent** *(significant relationship exists)*

In [1]:
# Importing required libraries
import numpy as np
import seaborn as sns
import pandas as pd
from scipy.stats import chi2_contingency

## 📦 Step 1 — Load the Titanic Dataset

We use Seaborn's built-in Titanic dataset — a real-world dataset containing information about 891 Titanic passengers.

In [2]:
df = sns.load_dataset('titanic')

print(f"Dataset Shape : {df.shape}")
print(f"Columns       : {list(df.columns)}")
print("-" * 50)
df.head()

Dataset Shape : (891, 15)
Columns       : ['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare', 'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town', 'alive', 'alone']
--------------------------------------------------


,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True


## 📋 Step 2 — Build the Contingency Table

A **Contingency Table** (also called a Cross-Tab) shows the **frequency distribution** of two categorical variables together.

Here we cross-tabulate **Sex** vs **Survived** to see how many males/females survived or didn't survive.

In [3]:
contingency_table = pd.crosstab(df['sex'], df['survived'])

# Rename columns for clarity
contingency_table.columns = ['Did Not Survive (0)', 'Survived (1)']

print("-" * 45)
print("       CONTINGENCY TABLE")
print("-" * 45)
print(contingency_table)
print("=" * 45)
print(f"\nTotal Passengers : {contingency_table.values.sum()}")
print(f"Female Survival Rate : {233/(81+233)*100:.1f}%")
print(f"Male Survival Rate   : {109/(468+109)*100:.1f}%")

---------------------------------------------
       CONTINGENCY TABLE
---------------------------------------------
        Did Not Survive (0)  Survived (1)
sex                                      
female                   81           233
male                    468           109

Total Passengers : 891
Female Survival Rate : 74.2%
Male Survival Rate   : 18.9%


## 🔍 Step 3 — Perform the Chi-Square Test

`chi2_contingency()` returns 4 values:

| Output | Meaning |
|---|---|
| `chi2` | Chi-Square test statistic |
| `p_value` | Probability value |
| `dof` | Degrees of Freedom = (rows-1) × (cols-1) |
| `expected` | Expected frequencies if H₀ were true |

> **Expected Frequencies** tell us what the counts *would look like* if gender and survival were completely independent of each other.

In [4]:
chi2, p_value, dof, expected = chi2_contingency(contingency_table)

print(f"Chi-Square Statistic : {chi2:.4f}")
print(f"P-Value              : {p_value:.10f}")
print(f"Degrees of Freedom   : {dof}")
print("\nExpected Frequencies (if H₀ were true):")
print(pd.DataFrame(
    expected,
    index=['female', 'male'],
    columns=['Did Not Survive (Expected)', 'Survived (Expected)']
).round(2))

Chi-Square Statistic : 260.7170
P-Value              : 0.0000000000
Degrees of Freedom   : 1

Expected Frequencies (if H₀ were true):
        Did Not Survive (Expected)  Survived (Expected)
female                      193.47               120.53
male                        355.53               221.47


## 📊 Step 4 — Observed vs Expected Comparison

Let's compare what we actually observed vs what we would expect if there was NO relationship between gender and survival.

| | Did Not Survive | Survived |
|---|---|---|
| **Female (Observed)** | 81 | 233 |
| **Female (Expected)** | ~193 | ~121 |
| **Male (Observed)** | 468 | 109 |
| **Male (Expected)** | ~356 | ~221 |

> 💡 The **huge gap** between Observed and Expected values already tells us something significant is going on — women survived much more than expected by chance!

## ✅ Step 5 — Decision: Accept or Reject Null Hypothesis?

### Decision Rule:

| Condition | Decision |
|---|---|
| p_value < α (0.05) | ❌ Reject H₀ — relationship EXISTS |
| p_value ≥ α (0.05) | ✅ Accept H₀ — no relationship |

In [5]:
alpha = 0.05

print(f"Chi-Square Statistic : {chi2:.4f}")
print(f"P-Value              : {p_value:.10f}")
print(f"Alpha                : {alpha}")
print("-" * 50)

if p_value < alpha:
    print("Reject the Null Hypothesis")
    print("→ There IS a significant relationship between Gender and Survival.")
else:
    print("Accept the Null Hypothesis")
    print("→ No significant relationship found between Gender and Survival.")

Chi-Square Statistic : 260.7170
P-Value              : 0.0000000000
Alpha                : 0.05
--------------------------------------------------
Reject the Null Hypothesis
→ There IS a significant relationship between Gender and Survival.


## 📝 Conclusion

| Metric | Value |
|---|---|
| Dataset | Titanic (891 passengers) |
| Variable 1 | Sex (male / female) |
| Variable 2 | Survived (0 / 1) |
| Chi-Square Statistic | 260.717 |
| Degrees of Freedom | 1 |
| P-Value | ~0.0000000001 |
| Alpha (α) | 0.05 |
| Decision | ❌ Reject H₀ |

---

**Interpretation:**  
With a **Chi-Square statistic of 260.72** and a **p-value practically equal to 0**, we strongly reject the null hypothesis.  
Gender and Survival on the Titanic are **NOT independent** — they are **strongly associated**.

> Women had a **74.2% survival rate** while men had only a **18.9% survival rate** — clearly the famous *"women and children first"* policy had a major statistical impact.

---

## 🔁 All Four Tests — Final Comparison

| Test | Data Type | Our Question | Decision |
|---|---|---|---|
| Z-Test | Numerical | Height = 170cm? | ✅ Accept H₀ |
| One Sample T-Test | Numerical | Height = 170cm? | ✅ Accept H₀ |
| Two Sample T-Test | Numerical | Group A = Group B? | ❌ Reject H₀ |
| **Chi-Square Test** | **Categorical** | **Gender affects Survival?** | **❌ Reject H₀** |

> 💡 **Key Takeaway:** Chi-Square Test is your go-to tool whenever you want to find if two **categorical variables** are related. The higher the Chi-Square value and lower the p-value, the stronger the association.

